confirm.c

```diff
@@ -395,38 +398,102 @@ static struct confirm_query * confirm_pkt_parse4(const struct packet *pkt)/*{{{*
-		data = ntohs(pkt->icmpv6->seq);
-	} else if(pkt->icmpv6->icmp_type == ICMP6_TIMXCEED) {
-		if(pkt->icmpv6->icmp_code != ICMP_TIMXCEED_INTRANS) return NULL;
-		struct libnet_ipv6_hdr *rip;
-		struct libnet_icmpv6_hdr *ricmp;
-		rip = (struct libnet_ipv6_hdr *)(pkt->payload);
-		ricmp = (struct libnet_icmpv6_hdr *)(pkt->payload + LIBNET_IPV6_H);
-		memcpy(&dst.sin6_addr, &rip->ip_dst, sizeof(dst.sin6_addr));
-		uint32_t flags = *(uint32_t *)(rip->ip_flags);
-		traffic_class = (flags & 0x0FF00000) >> 20;
-		flow_label = (flags & 0x000FFFFF);
-		icmpid = ntohs(ricmp->id);
-		data = ntohs(ricmp->seq);
+		uint32_t tcp_sequence_number = ntohl(pkt->tcp->th_ack) - 1;
+		icmpid = (uint16_t) ((tcp_sequence_number >> 16) & 0x0000FFFF);
+		data = (uint16_t) (tcp_sequence_number & 0x0000FFFF);
+		logd(LOG_DEBUG, "%s TCP SYN-ACK received\n", __func__);
+	}
+	else if(pkt->ipv6->ip_nh==IPPROTO_ICMP6) {
+		if(pkt->icmpv6->icmp_type == ICMP6_ECHOREPLY) {
+			probe_type = 1;
+			// Echo reply
+			memcpy(&dst.sin6_addr, &pkt->ipv6->ip_src, sizeof(dst.sin6_addr));
+			icmpid = ntohs(pkt->icmpv6->id);
+			data = ntohs(pkt->icmpv6->seq);
+		}
+		else if(pkt->icmpv6->icmp_type == ICMP6_TIMXCEED) {
+			logd(LOG_DEBUG, "%s ICMP time exceeded received\n", __func__);
+
+			// Time exceeded
+			if(pkt->icmpv6->icmp_code != ICMP_TIMXCEED_INTRANS) return NULL;
+			struct libnet_ipv6_hdr *rip;
+			struct libnet_icmpv6_hdr *ricmp;
+			struct libnet_tcp_hdr *rtcp;
+			rip = (struct libnet_ipv6_hdr *)(pkt->payload);
+
+			if(rip->ip_nh==IPPROTO_ICMP6){
+				// Pacote interno ICMP
+				probe_type = 1;
+				ricmp = (struct libnet_icmpv6_hdr *)
+						(pkt->payload + LIBNET_IPV6_H);
+				memcpy(&dst.sin6_addr, &rip->ip_dst, sizeof(dst.sin6_addr));
+				uint32_t flags = *(uint32_t *)(rip->ip_flags);
+				traffic_class = (flags & 0x0FF00000) >> 20;
+				flow_label = (flags & 0x000FFFFF);
+				icmpid = ntohs(ricmp->id);
+				data = ntohs(ricmp->seq);
+			}
+			else if(rip->ip_nh==IPPROTO_TCP) {
+				// Pacote interno TCP
+				// TODO (rafael): draft
+				probe_type = 2;
+				rtcp = (struct libnet_tcp_hdr *)(pkt->payload + LIBNET_IPV6_H);
+				memcpy(&dst.sin6_addr, &rip->ip_dst, sizeof(dst.sin6_addr));
+				uint32_t flags = *(uint32_t *)(rip->ip_flags);
+				traffic_class = (flags & 0x0FF00000) >> 20;
+				flow_label = (flags & 0x000FFFFF);
+				data = (uint16_t) (ntohl(rtcp->th_seq) & 0x0000FFFF);
+			}
+			else {
+				logd(LOG_FATAL, "%s %d: no internal TCP or ICMPv6\n", __FILE__,
+					__LINE__);
+				return NULL;
+			}
+		}
+		else if((pkt->icmpv6->icmp_type == ICMP6_DST_UNREACH) &&
+			(pkt->icmpv6->icmp_code == ICMP6_DST_UNREACH_NOPORT)){
+			// Port unreachable
+			// TODO (rafael): draft (to be tested)
+			probe_type = 2;
+			struct libnet_ipv6_hdr *rip = (struct libnet_ipv6_hdr *)(pkt->payload);
+			struct libnet_tcp_hdr *rtcp = (struct libnet_tcp_hdr *)
+					(pkt->payload + LIBNET_IPV6_H);
+			memcpy(&dst.sin6_addr, &rip->ip_dst, sizeof(dst.sin6_addr));
```

Comment:
don't we need an `if` here to check whether `rip` is ICMP or TCP? i understand that only TCP can elicit `UNREACH_NOPORT`, but ICMP may elicit `UNREACH` responses.